### PR TITLE
Always trigger componentWillUnmount in StrictMode

### PR DESCRIPTION
**Pull request by tyao1**
## Summary

In StrictMode, React currently only triggers `componentWillUnmount` if `componentDidMount`  is defined.  This would miss detecting issues like initializing resources in constructor or componentWillMount, for example:
```
class Component {
  constructor() {
     this._subscriptions = new Subscriptions();
  }
  componentWillUnmount() {
     this._subscriptions.reset();
  } 
}
``` 

The PR makes `componentWillUnmount` always run in StrictMode.

## How did you test this change?

`yarn test ci`

### CODE REVIEW

File: packages/react-reconciler/src/ReactFiberCommitWork.js [hunk 1]
```diff
@@ -4572,10 +4572,12 @@ function invokeLayoutEffectMountInDEV(fiber: Fiber): void {
       }
       case ClassComponent: {
         const instance = fiber.stateNode;
-        try {
-          instance.componentDidMount();
-        } catch (error) {
-          captureCommitPhaseError(fiber, fiber.return, error);
+        if (typeof instance.componentDidMount === 'function') {
```
Comment (sammy-SC) [inline packages/react-reconciler/src/ReactFiberCommitWork.js:4575]: why is this needed?

**Review by sammy-SC — APPROVED** [review]

File: packages/react-reconciler/src/ReactFiberCommitWork.js (hunk 1 above)
Comment (tyao1) [reply packages/react-reconciler/src/ReactFiberCommitWork.js:4575]: The PR always sets `MountLayoutDev` on mount regardless if `instance.componentDidMount` exists, and when the flag is set, this code path always runs.
